text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,371 @@
+
+For functions with generic arguments, the vtable (not the trait object, only its vtable) is passed as an invisible argument on the stack/registers, which is present only in the morphomization of the function which takes a trait object in its generics. If a function has more than one `?Sized` generic argument, the morphomization (generated at compile time) which requires multiple trait objects should take multiple invisible vtable arguments. The arguments are automatically filled out when the callee puts  the trait object(s) into the angle brackets (if they don't provide all the trait objects required and fill some with `dyn Trait`, an error is produced as described back in the guide-level section).
+
+For types with generic arguments, the vtables exist inside instances of that type as invisible private fields. Like the vtable function arguments, these are only visible to the compiler (but they obviously contribute to the size, which remains known since vtable pointers are just pointers). If an associated function is called, however (no `self` — no access to the invisible vtable fields), the vtables should be provided as invisible vtable arguments, as with regular functions.
+
+## Edge cases
+None unresolved yet.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None known yet.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The constantly mentioned alternative is declaring the associated functions in traits which are meant to be trait objects with a `&self` parameter, effectively converting all associated functions into trait methods. This becomes a major source of confusion if the trait object is mutably borrowed elsewhere while this "`&self`ified" method is called. This introduces lots of unnecessary artificial curly-bracket scopes which hurt readability in most cases.
+
+That workaround is also wrong from a semantic standpoint, since trait methods by design should use the fields of the trait object in some way, either directly or indirectly. Again, an unused `&self` just to combat language limitations introduces confusion.
+
+Last but not least, "`&self`ifying" works terribly when the trait and the code creating and using trait objects of it are in different crates made by different people. In this case, the library developer has to add `&self` to the associated trait functions, hurting the code using the trait without trait object, which loses the ability to use its associated functions without an actual object of that trait. As a result, library developers would be forced to have the associated functions in their struct's main `impl` block and "`&self`ified" wrappers around these in trait implementations. This is incomprehensibly hacky and non-idiomatic, seriously hurting library design in the long run but inapparent in the beginning of development.
```

Comment:
You can always convert a trait into one that can be used as a trait object. 

```rust
// lib a
trait Foo {
    fn foo() {}
    // ...
}
```

```rust
// lib b
trait DynFoo {
    fn dyn_foo(&self) {}
    // ...
}

impl<T: Foo + ?Sized> DynFoo for T {
    fn dyn_foo(&self) { T::foo() }
}
```

Then, whenever you need dynamic dispatch you use `DynFoo`, and this can be seemlessly integrated into an existing code base.

Reply:
> That workaround is also wrong from a semantic standpoint, since trait methods by design should use the fields of the trait object in some way, either directly or indirectly. Again, an unused `&self` just to combat language limitations introduces confusion.

As mentioned by the RFC, the "newtrait" solution is not less ugly. Why create a macro-worthy construct if you can implement the same thing idiomatically with a little bit of new syntax?